### PR TITLE
Added n_restarts as hyperparameter for CosineAnnealing

### PR DESCRIPTION
**Pull request by ravinkohli**

### CODE REVIEW

File: autoPyTorch/pipeline/components/training/trainer/base_trainer.py [hunk 1]
```diff
@@ -260,6 +260,11 @@ def prepare(
         if self.use_snapshot_ensemble:
             self.model_snapshots = list()
 
+        # in case we are using, swa or se with early stopping,
+        # we need to make sure network params are only updated
+        # from the swa model if the swa model was actually updated
+        self.swa_updated: bool = False
+
```
Comment (nabenabe0928) [inline autoPyTorch/pipeline/components/training/trainer/base_trainer.py:267]: It seems no description about what SWA (stochastic weighed average?) means, so could you add the description in the documentation of this class?

File: autoPyTorch/pipeline/components/training/trainer/base_trainer_choice.py [hunk 2]
```diff
@@ -410,7 +410,7 @@ def _fit(self, X: Dict[str, Any], y: Any = None, **kwargs: Any) -> torch.nn.Modu
             if 'cuda' in X['device']:
                 torch.cuda.empty_cache()
 
-        if self.choice.use_stochastic_weight_averaging:
+        if self.choice.use_stochastic_weight_averaging and self.choice.swa_updated:
```
Comment (nabenabe0928) [inline autoPyTorch/pipeline/components/training/trainer/base_trainer_choice.py:413]: When do you have to set `self.choice.swa_updated` back to `False`?
We need this change, right?

**Review by nabenabe0928 — CHANGES_REQUESTED** [review]

File: autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py [hunk 3]
```diff
@@ -69,13 +72,12 @@ def get_properties(dataset_properties: Optional[Dict[str, Any]] = None) -> Dict[
 
     @staticmethod
     def get_hyperparameter_search_space(dataset_properties: Optional[Dict] = None,
-                                        T_0: Tuple[Tuple[int, int], int] = ((1, 20), 1),
-                                        T_mult: Tuple[Tuple[float, float], float] = ((1.0, 2.0), 1.0)
+                                        n_restarts: Tuple[Tuple, int] = ((1, 6), 3)
```
[the remark this answers is not in the document]
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py:75]: yes, we need to keep it this way as its a part of how we are applying the search space update. 
This tuple is passed by the HyperparameterSearchSpaceUpdates

[the remark this answers is not in the document]
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py:75]: I checked `HyperparameterSearchSpaceUpdates`, but none of the function has returns. 

In any case, it seems this is not package dependency, so we should fix it in the near future, because meta-programming is not intuitive and easy to understand. (e.g. what is T_0[0][0] here even though we use T_0 as an integer in other locations. The name should have some kind of consistency.)

[the remark this answers is not in the document]
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py:75]: Yeah I agree, can you raise an issue for this?

File: autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py [hunk 4]
```diff
@@ -52,10 +51,14 @@ def fit(self, X: Dict[str, Any], y: Any = None) -> BaseLRComponent:
         # Make sure there is an optimizer
         self.check_requirements(X, y)
 
+        # initialise required attributes for the scheduler
+        T_mult: int = 1
+        T_0: int = min(math.floor(X['epochs'] / self.n_restarts), 1)
```
[the remark this answers is not in the document]
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py:56]: Yes it was a typo :P. I have used T_0, T_mult = (20, 1) as to what @ArlindKadra suggested. The problem this solves that in our case, we are using Hyperband and with that sometimes we have no. of epochs even less than the set value for T_0 and hence for some fidelity(Value of epochs) there never is a restart. With the above solution, we can guarantee that restarts will happen and the number of restarts is set to the value of the hyperparameter. Ofcourse, this is my idea as to how this is helping. But I am open to discuss for any improvements or changes needed.

File: autoPyTorch/pipeline/components/training/trainer/base_trainer.py (hunk 1 above)
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/training/trainer/base_trainer.py:267]: Sure

File: autoPyTorch/pipeline/components/training/trainer/base_trainer_choice.py (hunk 2 above)
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/training/trainer/base_trainer_choice.py:413]: I think for some pipeline it's enough that we change this once to True so to make sure the swa_model has meaning full weights. Especially in early stopping, there is a scenario when swa parameters were not updated. Leading to the model getting nearly random weights.

File: autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py [hunk 5]
```diff
@@ -26,14 +27,12 @@ class CosineAnnealingWarmRestarts(BaseLRComponent):
```
[the remark this answers is not in the document]
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py:27]: Sure

File: autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py:75]: Okay, I will do that, thanks:)

File: autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py]: Ok, I got what you mean. Probably, we should discuss it more closely.
I will just put my opinion and open questions for later discussion.

1. The combination of Hyperband and warm restart seems not good (because cyclic scheduling weakens the correlation of performance in terms of various time budgets.) 
2. When you think about the objective function as `f(λ, B)` where `λ` is a configuration and `B` is a budget, preferably `f(λ, B) ≒ f(λ, B + 1)`. However, in this case, the length of 1 cycle depends on `B` (, which is not a hyperparameter), we cannot measure the accurate performance.

File: autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ArlindKadra) [reply autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py:75]: Hey @ravinkohli, can we change `n_restarts: Tuple[Tuple, int]` to `n_restarts: Tuple[Tuple[int, int], int]`, so we can we consistent with the other modules ?

[the remark this answers is not in the document]
Comment (ArlindKadra) [reply autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py:75]: Could we also keep the `T_mul` to 2 and calculate the initial budget `T_0`. If I am not mistaken you can use the formula for the sum of a geometric progression. I think like this it would converge better.

**Review by ArlindKadra — CHANGES_REQUESTED** [review]
Left a few comments, but overall the PR is good :).

File: autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ArlindKadra) [reply autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py]: @nabenabe0928 

- Right on the first point, which is why we thought of switching to the number of restarts. So, even if we could find a good `T_0` and `T_mul` it will not necessarily be good or transferable to the other fidelities because quite probably, the larger fidelity will not finish at a local minimum with those values. The number of restarts however is different, saying that 2 restarts work for a lower fidelity, makes it more transferable to a higher one and it will converge to a nicer region.
- I proposed a better solution based on the budget given to calculate the initial `T_0` given a fixed `T_mult` of 2.0, in case you and @ravinkohli also like it.
- Disagree that the original code was more accurate in multi-fidelity optimization. I would argue that the original code was worse, as you want the good hyperparameter configurations on lower fidelities, to be the same good ones on higher ones. Just the different `T_mul `and` T_0` diverge the configurations between the fidelities, which does not help with the correlation.

[the remark this answers is not in the document]
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py]: @ArlindKadra 
For the third point, I did not say the original code is better, but I said the original code is more accurate in terms of the fidelity measure. Your statement is not mathematically clear, so please elaborate your opinion a bit more. 
I have to mention that Budget should not be a control parameter of hyperparameters.

[the remark this answers is not in the document]
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py]: About the third point, I did not say the original code is better, but I said the original code is more accurate in terms of the fidelity measure.

I do not know how familiar you are with the successive halving theory, but we need to maintain `f(x, B) ≒ f(x, B)`.
However, in this case, the following two cases will show the quite results:
(1). Train a model with config `x` for 50 epochs (out of 100 epochs, it means we cut off at the 50 epoch).
(2). Train a model with config `x` for 50 epochs (out of 50 epochs, it means we complete a full training).
I mentioned this really matters.

[the remark this answers is not in the document]
Comment (ArlindKadra) [reply autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py]: With f(x, B) ≒ f(x, B) you probably mean f(x, B) ≒ f(x, B + 1).
If so, in your opinion, what would be a closer approximate performance between 2 different budgets/fidelities, a certain T_mul and T_0 or the number of restarts?

[the remark this answers is not in the document]
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/setup/lr_scheduler/CosineAnnealingWarmRestarts.py]: I agree with you that there might be a loss of correlation. I think as this is being merged into the regularisation cocktails, when we merge this to the refactor_development branch we can have a discussion about it

**Review by ArlindKadra — APPROVED** [review]
